### PR TITLE
Kungfuflex/runeid leb128

### PR DESCRIPTION
**Pull request by hudeintres**

### CODE REVIEW

File: assembly/indexer/protomessage/MessageContext.ts [hunk 1]
```diff
@@ -86,8 +83,8 @@ export class MessageContext {
       this.runes.push(rune);
     }
   }
-  initializeProtocol(): u128 {
-    const tag = this.protocolTag();
+  initializeProtocol<T extends MessageContext>(): u128 {
```
Comment (clothic) [inline assembly/indexer/protomessage/MessageContext.ts:86]: whats the point of this? to allow the protocol tag to be overridden?

File: assembly/indexer/ProtoruneRuneId.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      let size = readULEB128ToU128(input, current);
+      if (size === usize.MAX_VALUE) return changetype<Array<u128>>(0); //can choose to continue or return
+      input.shrinkFront(size);
+      result.push(current);
+    }
+    return result;
+  }
+  // ary should have the string identifier for the protorune first, then the data that should be encoded
+  // Ex: ["OYL", RuneBlock1, RuneTx1, RuneBlock2, RuneTx2]. Guarantees that the padded bits can be ignored and the first data byte is nonzero (must be a string)
+  // Important note: collisions with existing RuneIds are not explicitly handled. Authors of subprotocols should ensure that the ProtoruneRuneId scheme does not conflict with existing rune ids.
```
Comment (clothic) [inline assembly/indexer/ProtoruneRuneId.ts:26]: hm, if we can't guarantee that protorunrRuneIds are unique, then there could be attacks on subprotocols.

For example, if oylswap mints a protorune with ProtoruneRuneId as an LP token, someone can just copy the params called here to mint their own protorune that will be indistinguishable from the real one

File: assembly/indexer/ProtoruneRuneId.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  // Do note that a collision would be extremely unlikely: it would require 14 bytes of pure 0s input as data, in order for txIndex to be valid, and then the data needs to be higher than 840,000 for the block.
+  static encode(ary: Array<u128>): ProtoruneRuneId {
+    const data = Box.concat(
+      ary.map<Box>((v: u128, i: i32, ary: Array<u128>) =>
+        Box.from(toLEB128(v)),
+      ),
+    );
+    const result = new ProtoruneRuneId(0, 0);
+    if (data.byteLength > 32) return result;
```
[the remark this answers is not in the document]
Comment (hudeintres) [reply assembly/indexer/ProtoruneRuneId.ts:35]: Yes

File: assembly/indexer/ProtoruneRuneId.ts (hunk 2 above)
Comment (hudeintres) [reply assembly/indexer/ProtoruneRuneId.ts:26]: As discussed in chat, this should be implemented in the subprotocol

**Pull request merged**